multirun/multirun.go

```diff
@@ -10,49 +10,99 @@ type multirun struct {
 	commands               []command
 	stdoutSink, stderrSink io.Writer
 	args                   []string
-	parallel               bool
+	jobs                   int
 	quiet                  bool
 }
 
 func (m multirun) run(ctx context.Context) error {
-	if m.parallel {
-		errs := make(chan error)
-		for _, cmd := range m.commands {
-			p := process{
-				tag:        cmd.Tag,
-				path:       cmd.Path,
-				stdoutSink: m.stdoutSink,
-				stderrSink: m.stderrSink,
-				args:       m.args,
-				addTag:     true,
-			}
-			go func() {
-				errs <- p.run(ctx)
-			}()
+	// minimum concurrency is 1, which is sequential
+	if m.jobs < 0 {
+		return fmt.Errorf("jobs must be at least 0")
+	}
+
+	if m.jobs == 0 {
+		return m.unboundedExecution(ctx)
+	}
+
+	return m.boundedExecution(ctx)
+}
+
+// unboundedExecution execute multiple commands without concurrency limit
+func (m multirun) unboundedExecution(ctx context.Context) error {
+	errs := make(chan error)
+
+	for _, cmd := range m.commands {
+		p := process{
+			tag:        cmd.Tag,
+			path:       cmd.Path,
+			stdoutSink: m.stdoutSink,
+			stderrSink: m.stderrSink,
+			args:       m.args,
+			addTag:     true,
 		}
-		var firstError error
-		for range m.commands {
-			if err := <-errs; firstError == nil {
-				firstError = err
-			}
+
+		go func() {
+			errs <- p.run(ctx)
+		}()
+	}
+
+	var firstError error
+	for range m.commands {
+		if err := <-errs; firstError == nil {
+			firstError = err
 		}
-		return firstError
-	} else {
-		for _, cmd := range m.commands {
-			p := process{
-				tag:        cmd.Tag,
-				path:       cmd.Path,
-				stdoutSink: m.stdoutSink,
-				stderrSink: m.stderrSink,
-				args:       m.args,
-			}
-			if !m.quiet {
-				fmt.Fprintf(m.stderrSink, "Running %s\n", cmd.Tag)
-			}
-			if err := p.run(ctx); err != nil {
-				return err
-			}
+	}
+
+	return firstError
+}
+
+// boundedExecution execute multiple commands using a sized worker pool
+func (m multirun) boundedExecution(ctx context.Context) error {
+	// errs should be buffered to avoid blocking
+	// when len(m.commands) > m.jobs
+	errs := make(chan error, m.jobs)
```

Comment:
I doubt this compiles :)

```go
errs := make(chan error, len(m.jobs))
```

Reply:
I think you are thinking `len(m.commands)` which is a slice?

`m.jobs` is int value which is the number of workers

The code definitely does compile currently 🤓 

---

Perhaps I can do something like `max(len(m.commands), m.jobs)`? But may be thats overkilled? 🤔

Reply:
Yeah, after re-reading the code, it seems like we need to wait for the `commands <- cmd` loop to finish fully before we start consuming the `errs` channel. That means the `errs` channel need to hold as much as the size of that loop (which is `len(m.commands)`).

Let me fix it to that

Reply:
Oh, yeah, I got confused. But you are right, the channel buffer size needs to match commands. Thanks!